src/rust/src/wallet_scanner.rs

```diff
@@ -269,7 +273,8 @@ impl<D: Domain> DynamicUsage for OutputReplier<D> {
 }
 
 /// A batch of outputs to trial decrypt.
-struct Batch<D: BatchDomain, Output: ShieldedOutput<D, ENC_CIPHERTEXT_SIZE>> {
+struct Batch<A, D: BatchDomain, Output: ShieldedOutput<D, ENC_CIPHERTEXT_SIZE>> {
+    tags: Vec<A>,
     ivks: Vec<D::IncomingViewingKey>,
```

Comment:
There are two bugs here:
- `tags` is not included in the memory usage calculations.
- `ivks` now _does_ allocate heap memory, which is not included in the memory usage calculations.

I think we should just fix this properly by implementing `memuse::DynamicUsage` on the relevant types so we can rely on the relevant bounds. I'll go make the requisite PRs.

Reply:
I fixed the first of these bugs.

Reply:
I've opened https://github.com/zcash/librustzcash/pull/652 which makes the relevant changes to that copy of `BatchRunner`. The first commit of that PR is equivalent to the commit you pushed to fix the first of the above bugs. If you're happy with the approach I've taken in this PR, I'll port its commits onto this PR as well.